### PR TITLE
STITCH-1989 (1987/2207/1988/1985) Multi-User

### PR DESCRIPTION
**Pull request by jsflax**
Also includes drive-bys to fix evg.

### CODE REVIEW

File: core/sdk/src/test/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuthUnitTests.java [hunk 1]
```diff
@@ -17,14 +17,16 @@
 package com.mongodb.stitch.core.auth.internal;
```
Comment (adamchel) [inline core/sdk/src/test/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuthUnitTests.java]: [task] can you also add tests where you:
* log back in to a logged out (but not removed) user
* log back in to a removed user

File: core/sdk/src/test/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuthUnitTests.java [hunk 2]
```diff
@@ -196,22 +201,105 @@ public void testLogoutBlocking() {
...
+    final CoreStitchUser user3 =
+        auth.loginWithCredentialInternal(new UserPasswordCredential("bye", "there"));
 
+    assertEquals(auth.listUsers().getLast(), user3);
     assertTrue(auth.isLoggedIn());
 
     auth.logoutInternal();
 
+    // assert that though one user is logged out, two users are still listed
+    assertEquals(auth.listUsers().size(), 3);
```
Comment (adamchel) [inline core/sdk/src/test/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuthUnitTests.java]: [task] can you add a check that the profile information is still visible for the logged out user?

File: core/sdk/src/test/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuthUnitTests.java [hunk 3]
```diff
@@ -541,8 +630,8 @@ public void testProfileRequestFailureEdgeCases() {
       // do nothing
     }
 
-    assertFalse(auth.isLoggedIn());
-    assertNull(auth.getUser());
+    assertTrue(auth.isLoggedIn());
+    assertNotNull(auth.getUser());
```
Comment (adamchel) [inline core/sdk/src/test/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuthUnitTests.java]: [task] verify that the ID of this user is the ID of the original user

File: server/core/src/main/java/com/mongodb/stitch/server/core/auth/StitchAuth.java [hunk 4]
```diff
@@ -73,10 +98,27 @@
...
   @Nullable
   StitchUser getUser();
+
+
+  /**
+   * Returns a set of all logged in users.
+   *
+   * @return the set of currently logged in users
+   */
+  LinkedList<StitchUser> listUsers();
```
Comment (adamchel) [inline server/core/src/main/java/com/mongodb/stitch/server/core/auth/StitchAuth.java]: [nit] ditto that this should be `List`

File: core/sdk/src/main/java/com/mongodb/stitch/core/auth/internal/AuthInfo.java [hunk 5]
```diff
@@ -71,30 +76,68 @@ static AuthInfo readFromApi(final InputStream is) throws IOException {
...
+    final String rawInfo = storage.get(ACTIVE_USER_STORAGE_NAME);
     if (rawInfo == null) {
       return null;
     }
+
     return StitchObjectMapper.getInstance().readValue(rawInfo, StoreAuthInfo.class);
   }
 
-  void writeToStorage(final Storage storage) throws IOException {
+  static LinkedList<AuthInfo> readCurrentUsersFromStorage(
```
[the remark this answers is not in the document]
Comment (jsflax) [reply core/sdk/src/main/java/com/mongodb/stitch/core/auth/internal/AuthInfo.java]: Yeah I agree with making the user exposed API `List`s. The linked lists provide us with some nice ordering for users, and is being used as a ringbuffer for logged out but not removed users.

File: .evg.yml [hunk 6]
```diff
@@ -59,7 +59,7 @@ functions:
                   export ANDROID_SDK_ROOT=$SDK_HOME
                   export ANDROID_SDK_HOME=$SDK_HOME
 
-                  SDK_PACKAGE=sdk-tools-linux-4333796.zip
+                  SDK_PACKAGE=sdk-tools-darwin-4333796.zip
```
[the remark this answers is not in the document]
Comment (jsflax) [reply .evg.yml]: I was having trouble getting it to work on the linux boxes– can try again though.

File: android/coretest/src/androidTest/java/com/mongodb/stitch/android/core/StitchAppClientIntTests.kt [hunk 7]
```diff
@@ -127,8 +127,69 @@ class StitchAppClientIntTests : BaseStitchAndroidIntTest() {
         assertTrue(client.auth.isLoggedIn)
         assertEquals(client.auth.user!!.loggedInProviderType, UserPasswordAuthProvider.TYPE)
 
-        // Verify that logout clears storage
-        Tasks.await(client.auth.logout())
+        // check everything is as it was
```
[the remark this answers is not in the document]
Comment (jsflax) [reply android/coretest/src/androidTest/java/com/mongodb/stitch/android/core/StitchAppClientIntTests.kt]: I also do this slightly later in this test.

**Comment by jsflax** [comment]
@adamchel PTAL

File: .evg.yml [hunk 8]
```diff
@@ -369,16 +399,16 @@ post:
...
+    mongodb_url: https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-4.0.2.tgz
+    transpiler_target: node8-linux
   tasks:
-    # - name: run_core_tests
+    - name: run_core_tests
     - name: run_android_tests
-    # - name: lint
-    # - name: finalize_coverage
+    - name: lint
+    - name: finalize_coverage
```
Comment (adamchel) [inline .evg.yml]: [nit] newline

File: core/sdk/src/test/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuthUnitTests.java [hunk 9]
```diff
@@ -631,6 +659,7 @@ public void testProfileRequestFailureEdgeCases() {
     }
 
     assertTrue(auth.isLoggedIn());
+    assertEquals(auth.getUser().getId(), user1.getId());
     assertNotNull(auth.getUser());
```
Comment (adamchel) [inline core/sdk/src/test/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuthUnitTests.java]: [nit] this check is redundant now

**Pull request merged**